### PR TITLE
[SYCL] Replace __pipe keyword with pipe attribute

### PR DESCRIPTION
**Pull request by vmaksimo**

### CODE REVIEW

File: clang/include/clang/Basic/AttrDocs.td [hunk 1]
```diff
@@ -1872,6 +1872,16 @@ function, and no effect otherwise.
   }];
 }
 
+def IntelFPGAPipeDocs : Documentation {
+  let Category = DocCatStmt;
+  let Heading = "pipe (read_only, write_only)";
+  let Content = [{
+This attribute applies to a type to indicate that it's is a pipe type.
```
Comment (sergey-semenov) [inline clang/include/clang/Basic/AttrDocs.td]:
```suggestion
This attribute applies to a type to indicate that it is a pipe type.
```

Comment (sergey-semenov) [inline clang/include/clang/Basic/AttrDocs.td]:
```suggestion
This attribute applies to a type to indicate that it is a pipe type.
```

File: clang/include/clang/Basic/Attr.td [hunk 2]
```diff
@@ -1694,6 +1694,13 @@ def IntelFPGASimpleDualPort : Attr {
   let Documentation = [IntelFPGASimpleDualPortAttrDocs];
 }
 
+def IntelFPGAPipe : TypeAttr {
```
Comment (bader) [inline clang/include/clang/Basic/Attr.td]: Should we use "Intel" prefix here?
I understand that it's Intel extension, but I think Xilinx is also using this feature.
Can we use something more generic in the source code? E.g. `SYCLFPGAPipe`?

+@keryell, +@mkinsner

Comment (vmaksimo) [reply clang/include/clang/Basic/Attr.td]: Sure, let's use `SYCLFPGAPipe`

**Comment by MrSidims** [comment]
> It is not clear to me why there is a need for an attribute.
> If I understand correctly the pipes are represented with a special datatype and its member functions are mangled accordingly.
> In the LLVM lowering to SPIR-V or whatever you can recognize the members functions and lower them or I miss something.

Your understanding is correct. To answer the question, let me split it into two parts (please correct me misunderstood you):
1. Why do we need clang representation in clang? (basically, we could forward declare SPIR-V type, aka `class __spirv_ReadPipe;` and use it later)
2. Why do we add new attributes (instead of previously done keywords)?

So, even pipe built-ins are lowered into SPIR-V instructions, we can re-use clang pipe type. It will allow us to do special diagnostics for pipes in the future.
Replacing the keyword with the attribute is making code safer in case of external libraries linkage (well, "pipe" is kinda popular word in system programming). That is why it was declared not as "pipe", but as "__pipe" originally.

**Pull request merged**